File: tools/taos-tools/deps/avro/lang/java/avro/src/main/java/org/apache/avro/LogicalTypes.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avro;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class LogicalTypes {

  private static final Logger LOG = LoggerFactory.getLogger(LogicalTypes.class);

  public interface LogicalTypeFactory {
    LogicalType fromSchema(Schema schema);

    default String getTypeName() {
      throw new UnsupportedOperationException("LogicalTypeFactory TypeName has not been provided");
    }
  }

  private static final Map<String, LogicalTypeFactory> REGISTERED_TYPES = new ConcurrentHashMap<>();

  /**
   * Register a logical type.
   *
   * @param factory The logical type factory
   *
   * @throws NullPointerException if {@code factory} or
   *                              {@code factory.getTypedName()} is {@code null}
   */
  public static void register(LogicalTypeFactory factory) {
    Objects.requireNonNull(factory, "Logical type factory cannot be null");
    register(factory.getTypeName(), factory);
  }

  /**
   * Register a logical type.
   *
   * @param logicalTypeName The logical type name
   * @param factory         The logical type factory
   *
   * @throws NullPointerException if {@code logicalTypeName} or {@code factory} is
   *                              {@code null}
   */
  public static void register(String logicalTypeName, LogicalTypeFactory factory) {
    Objects.requireNonNull(logicalTypeName, "Logical type name cannot be null");
    Objects.requireNonNull(factory, "Logical type factory cannot be null");

    try {
      String factoryTypeName = factory.getTypeName();
      if (!logicalTypeName.equals(factoryTypeName)) {
        LOG.debug("Provided logicalTypeName '{}' does not match factory typeName '{}'", logicalTypeName,
            factoryTypeName);
      }
    } catch (UnsupportedOperationException ignore) {
      // Ignore exception, as the default interface method throws
      // UnsupportedOperationException.
    }

    REGISTERED_TYPES.put(logicalTypeName, factory);
  }

  /**
   * Return an unmodifiable map of any registered custom {@link LogicalType}
   */
  public static Map<String, LogicalTypes.LogicalTypeFactory> getCustomRegisteredTypes() {
    return Collections.unmodifiableMap(REGISTERED_TYPES);
  }

  /**
   * Returns the {@link LogicalType} from the schema, if one is present.
   */
  public static LogicalType fromSchema(Schema schema) {
    return fromSchemaImpl(schema, true);
  }

  public static LogicalType fromSchemaIgnoreInvalid(Schema schema) {
    return fromSchemaImpl(schema, false);
  }

  private static LogicalType fromSchemaImpl(Schema schema, boolean throwErrors) {
    final LogicalType logicalType;
    final String typeName = schema.getProp(LogicalType.LOGICAL_TYPE_PROP);

    if (typeName == null) {
      return null;
    }

    try {
      switch (typeName) {
      case TIMESTAMP_MILLIS:
        logicalType = TIMESTAMP_MILLIS_TYPE;
        break;
      case DECIMAL:
        logicalType = new Decimal(schema);
        break;
      case UUID:
        logicalType = UUID_TYPE;
        break;
      case DATE:
        logicalType = DATE_TYPE;
        break;
      case TIMESTAMP_MICROS:
        logicalType = TIMESTAMP_MICROS_TYPE;
        break;
      case TIME_MILLIS:
        logicalType = TIME_MILLIS_TYPE;
        break;
      case TIME_MICROS:
        logicalType = TIME_MICROS_TYPE;
        break;
      case LOCAL_TIMESTAMP_MICROS:
        logicalType = LOCAL_TIMESTAMP_MICROS_TYPE;
        break;
      case LOCAL_TIMESTAMP_MILLIS:
        logicalType = LOCAL_TIMESTAMP_MILLIS_TYPE;
        break;
      default:
        final LogicalTypeFactory typeFactory = REGISTERED_TYPES.get(typeName);
        logicalType = (typeFactory == null) ? null : typeFactory.fromSchema(schema);
        break;
      }

      // make sure the type is valid before returning it
      if (logicalType != null) {
        logicalType.validate(schema);
      }
    } catch (RuntimeException e) {
      LOG.debug("Invalid logical type found", e);
      if (throwErrors) {
        throw e;
      }
      LOG.warn("Ignoring invalid logical type for name: {}", typeName);
      // ignore invalid types
      return null;
    }

    return logicalType;
  }

  private static final String DECIMAL = "decimal";
  private static final String UUID = "uuid";
  private static final String DATE = "date";
  private static final String TIME_MILLIS = "time-millis";
  private static final String TIME_MICROS = "time-micros";
  private static final String TIMESTAMP_MILLIS = "timestamp-millis";
  private static final String TIMESTAMP_MICROS = "timestamp-micros";
  private static final String LOCAL_TIMESTAMP_MILLIS = "local-timestamp-millis";
  private static final String LOCAL_TIMESTAMP_MICROS = "local-timestamp-micros";

  /** Create a Decimal LogicalType with the given precision and scale 0 */
  public static Decimal decimal(int precision) {
    return decimal(precision, 0);
  }

  /** Create a Decimal LogicalType with the given precision and scale */
  public static Decimal decimal(int precision, int scale) {
    return new Decimal(precision, scale);
  }

  private static final LogicalType UUID_TYPE = new LogicalType("uuid");

  public static LogicalType uuid() {
    return UUID_TYPE;
  }

  private static final Date DATE_TYPE = new Date();

  public static Date date() {
    return DATE_TYPE;
  }

  private static final TimeMillis TIME_MILLIS_TYPE = new TimeMillis();

  public static TimeMillis timeMillis() {
    return TIME_MILLIS_TYPE;
  }

  private static final TimeMicros TIME_MICROS_TYPE = new TimeMicros();

  public static TimeMicros timeMicros() {
    return TIME_MICROS_TYPE;
  }

  private static final TimestampMillis TIMESTAMP_MILLIS_TYPE = new TimestampMillis();

  public static TimestampMillis timestampMillis() {
    return TIMESTAMP_MILLIS_TYPE;
  }

  private static final TimestampMicros TIMESTAMP_MICROS_TYPE = new TimestampMicros();

  public static TimestampMicros timestampMicros() {
    return TIMESTAMP_MICROS_TYPE;
  }

  private static final LocalTimestampMillis LOCAL_TIMESTAMP_MILLIS_TYPE = new LocalTimestampMillis();

  public static LocalTimestampMillis localTimestampMillis() {
    return LOCAL_TIMESTAMP_MILLIS_TYPE;
  }

  private static final LocalTimestampMicros LOCAL_TIMESTAMP_MICROS_TYPE = new LocalTimestampMicros();

  public static LocalTimestampMicros localTimestampMicros() {
    return LOCAL_TIMESTAMP_MICROS_TYPE;
  }

  /** Decimal represents arbitrary-precision fixed-scale decimal numbers */
  public static class Decimal extends LogicalType {
    private static final String PRECISION_PROP = "precision";
    private static final String SCALE_PROP = "scale";

    private final int precision;
    private final int scale;

    private Decimal(int precision, int scale) {
      super(DECIMAL);
      this.precision = precision;
      this.scale = scale;
    }

    private Decimal(Schema schema) {
      super("decimal");
      if (!hasProperty(schema, PRECISION_PROP)) {
        throw new IllegalArgumentException("Invalid decimal: missing precision");
      }

      this.precision = getInt(schema, PRECISION_PROP);

      if (hasProperty(schema, SCALE_PROP)) {
        this.scale = getInt(schema, SCALE_PROP);
      } else {
        this.scale = 0;
      }
    }

    @Override
    public Schema addToSchema(Schema schema) {
      super.addToSchema(schema);
      schema.addProp(PRECISION_PROP, precision);
      schema.addProp(SCALE_PROP, scale);
      return schema;
    }

    public int getPrecision() {
      return precision;
    }

    public int getScale() {
      return scale;
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      // validate the type
      if (schema.getType() != Schema.Type.FIXED && schema.getType() != Schema.Type.BYTES) {
        throw new IllegalArgumentException("Logical type decimal must be backed by fixed or bytes");
      }
      if (precision <= 0) {
        throw new IllegalArgumentException("Invalid decimal precision: " + precision + " (must be positive)");
      } else if (precision > maxPrecision(schema)) {
        throw new IllegalArgumentException("fixed(" + schema.getFixedSize() + ") cannot store " + precision
            + " digits (max " + maxPrecision(schema) + ")");
      }
      if (scale < 0) {
        throw new IllegalArgumentException("Invalid decimal scale: " + scale + " (must be positive)");
      } else if (scale > precision) {
        throw new IllegalArgumentException(
            "Invalid decimal scale: " + scale + " (greater than precision: " + precision + ")");
      }
    }

    private long maxPrecision(Schema schema) {
      if (schema.getType() == Schema.Type.BYTES) {
        // not bounded
        return Integer.MAX_VALUE;
      } else if (schema.getType() == Schema.Type.FIXED) {
        int size = schema.getFixedSize();
        return Math.round(Math.floor(Math.log10(2) * (8 * size - 1)));
      } else {
        // not valid for any other type
        return 0;
      }
    }

    private boolean hasProperty(Schema schema, String name) {
      return (schema.getObjectProp(name) != null);
    }

    private int getInt(Schema schema, String name) {
      Object obj = schema.getObjectProp(name);
      if (obj instanceof Integer) {
        return (Integer) obj;
      }
      throw new IllegalArgumentException(
          "Expected int " + name + ": " + (obj == null ? "null" : obj + ":" + obj.getClass().getSimpleName()));
    }

    @Override
    public boolean equals(Object o) {
      if (this == o)
        return true;
      if (o == null || getClass() != o.getClass())
        return false;

      Decimal decimal = (Decimal) o;

      if (precision != decimal.precision)
        return false;
      return scale == decimal.scale;
    }

    @Override
    public int hashCode() {
      int result = precision;
      result = 31 * result + scale;
      return result;
    }
  }

  /** Date represents a date without a time */
  public static class Date extends LogicalType {
    private Date() {
      super(DATE);
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      if (schema.getType() != Schema.Type.INT) {
        throw new IllegalArgumentException("Date can only be used with an underlying int type");
      }
    }
  }

  /** TimeMillis represents a time in milliseconds without a date */
  public static class TimeMillis extends LogicalType {
    private TimeMillis() {
      super(TIME_MILLIS);
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      if (schema.getType() != Schema.Type.INT) {
        throw new IllegalArgumentException("Time (millis) can only be used with an underlying int type");
      }
    }
  }

  /** TimeMicros represents a time in microseconds without a date */
  public static class TimeMicros extends LogicalType {
    private TimeMicros() {
      super(TIME_MICROS);
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      if (schema.getType() != Schema.Type.LONG) {
        throw new IllegalArgumentException("Time (micros) can only be used with an underlying long type");
      }
    }
  }

  /** TimestampMillis represents a date and time in milliseconds */
  public static class TimestampMillis extends LogicalType {
    private TimestampMillis() {
      super(TIMESTAMP_MILLIS);
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      if (schema.getType() != Schema.Type.LONG) {
        throw new IllegalArgumentException("Timestamp (millis) can only be used with an underlying long type");
      }
    }
  }

  /** TimestampMicros represents a date and time in microseconds */
  public static class TimestampMicros extends LogicalType {
    private TimestampMicros() {
      super(TIMESTAMP_MICROS);
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      if (schema.getType() != Schema.Type.LONG) {
        throw new IllegalArgumentException("Timestamp (micros) can only be used with an underlying long type");
      }
    }
  }

  public static class LocalTimestampMillis extends LogicalType {
    private LocalTimestampMillis() {
      super(LOCAL_TIMESTAMP_MILLIS);
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      if (schema.getType() != Schema.Type.LONG) {
        throw new IllegalArgumentException("Local timestamp (millis) can only be used with an underlying long type");
      }
    }
  }

  public static class LocalTimestampMicros extends LogicalType {
    private LocalTimestampMicros() {
      super(LOCAL_TIMESTAMP_MICROS);
    }

    @Override
    public void validate(Schema schema) {
      super.validate(schema);
      if (schema.getType() != Schema.Type.LONG) {
        throw new IllegalArgumentException("Local timestamp (micros) can only be used with an underlying long type");
      }
    }
  }

}

File: tools/taos-tools/deps/avro/lang/ruby/lib/avro/logical_types.rb
# -*- coding: utf-8 -*-
# frozen_string_literal: true
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'date'
require 'bigdecimal'
require 'bigdecimal/util'

module Avro
  module LogicalTypes
    ##
    # Base class for logical types requiring a schema to be present
    class LogicalTypeWithSchema
      ##
      # @return [Avro::Schema] The schema this logical type is dealing with
      attr_reader :schema

      ##
      # Build a new instance of a logical type using the provided schema
      #
      # @param schema [Avro::Schema]
      #     The schema to use with this instance
      #
      # @raise [ArgumentError]
      #     If the provided schema is nil
      def initialize(schema)
        raise ArgumentError, 'schema is required' if schema.nil?

        @schema = schema
      end

      ##
      # Encode the provided datum
      #
      # @param datum [Object] The datum to encode
      #
      # @raise [NotImplementedError]
      #     Subclass will need to override this method
      def encode(datum)
        raise NotImplementedError
      end

      ##
      # Decode the provided datum
      #
      # @param datum [Object] The datum to decode
      #
      # @raise [NotImplementedError]
      #     Subclass will need to override this method
      def decode(datum)
        raise NotImplementedError
      end
    end

    ##
    # Logical type to handle arbitrary-precision decimals using byte array.
    #
    # The byte array contains the two's-complement representation of the unscaled integer
    # value in big-endian byte order.
    class BytesDecimal < LogicalTypeWithSchema
      # Messages for exceptions
      ERROR_INSUFFICIENT_PRECISION = 'Precision is too small'
      ERROR_ROUNDING_NECESSARY     = 'Rounding necessary'
      ERROR_VALUE_MUST_BE_NUMERIC  = 'value must be numeric'

      # The pattern used to pack up the byte array (8 bit unsigned integer/char)
      PACK_UNSIGNED_CHARS = 'C*'

      # The number 10 as BigDecimal
      TEN = BigDecimal(10).freeze

      ##
      # @return [Integer] The number of total digits supported by the decimal
      attr_reader :precision

      ##
      # @return [Integer] The number of fractional digits
      attr_reader :scale

      ##
      # Build a new decimal logical type
      #
      # @param schema [Avro::Schema]
      #     The schema defining precision and scale for the conversion
      def initialize(schema)
        super

        @scale     = schema.scale.to_i
        @precision = schema.precision.to_i
        @factor    = TEN ** @scale
      end

      ##
      # Encode the provided value into a byte array
      #
      # @param value [BigDecimal, Float, Integer]
      #     The numeric value to encode
      #
      # @raise [ArgumentError]
      #     If the provided value is not a numeric type
      #
      # @raise [RangeError]
      #     If the provided value has a scale higher than the schema permits,
      #     or does not fit into the schema's precision
      def encode(value)
        raise ArgumentError, ERROR_VALUE_MUST_BE_NUMERIC unless value.is_a?(Numeric)

        to_byte_array(unscaled_value(value.to_d)).pack(PACK_UNSIGNED_CHARS).freeze
      end

      ##
      # Decode a byte array (in form of a string) into a BigDecimal of the
      # given precision and scale
      #
      # @param stream [String]
      #     The byte array to decode
      #
      # @return [BigDecimal]
      def decode(stream)
        from_byte_array(stream) / @factor
      end

      private

      ##
      # Convert the provided stream of bytes into the unscaled value
      #
      # @param stream [String]
      #     The stream of bytes to convert
      #
      # @return [Integer]
      def from_byte_array(stream)
        bytes    = stream.bytes
        positive = bytes.first[7].zero?
        total    = 0

        bytes.each_with_index do |value, ix|
          total += (positive ? value : (value ^ 0xff)) << (bytes.length - ix - 1) * 8
        end

        return total if positive

        -(total + 1)
      end

      ##
      # Convert the provided number into its two's complement representation
      # in network order (big endian).
      #
      # @param number [Integer]
      #     The number to convert
      #
      # @return [Array<Integer>]
      #     The byte array in network order
      def to_byte_array(number)
        [].tap do |result|
          loop do
            result.unshift(number & 0xff)
            number >>= 8

            break if (number == 0 || number == -1) && (result.first[7] == number[7])
          end
        end
      end

      ##
      # Get the unscaled value from a BigDecimal considering the schema's scale
      #
      # @param decimal [BigDecimal]
      #     The decimal to get the unscaled value from
      #
      # @return [Integer]
      def unscaled_value(decimal)
        details = decimal.split
        length  = details[1].length

        fractional_part = length - details[3]
        raise RangeError, ERROR_ROUNDING_NECESSARY if fractional_part > scale

        if length > precision || (length - fractional_part) > (precision - scale)
          raise RangeError, ERROR_INSUFFICIENT_PRECISION
        end

        (decimal * @factor).to_i
      end
    end

    module IntDate
      EPOCH_START = Date.new(1970, 1, 1)

      def self.encode(date)
        return date.to_i if date.is_a?(Numeric)

        (date - EPOCH_START).to_i
      end

      def self.decode(int)
        EPOCH_START + int
      end
    end

    module TimestampMillis
      def self.encode(value)
        return value.to_i if value.is_a?(Numeric)

        time = value.to_time
        time.to_i * 1000 + time.usec / 1000
      end

      def self.decode(int)
        s, ms = int / 1000, int % 1000
        Time.at(s, ms * 1000).utc
      end
    end

    module TimestampMicros
      def self.encode(value)
        return value.to_i if value.is_a?(Numeric)

        time = value.to_time
        time.to_i * 1000_000 + time.usec
      end

      def self.decode(int)
        s, us = int / 1000_000, int % 1000_000
        Time.at(s, us).utc
      end
    end

    module Identity
      def self.encode(datum)
        datum
      end

      def self.decode(datum)
        datum
      end
    end

    TYPES = {
      "bytes" => {
        "decimal" => BytesDecimal
      },
      "int" => {
        "date" => IntDate
      },
      "long" => {
        "timestamp-millis" => TimestampMillis,
        "timestamp-micros" => TimestampMicros
      },
    }.freeze

    def self.type_adapter(type, logical_type, schema = nil)
      return unless logical_type

      adapter = TYPES.fetch(type, {}.freeze).fetch(logical_type, Identity)
      adapter.is_a?(Class) ? adapter.new(schema) : adapter
    end
  end
end
